crates/rpc/rpc-eth-api/src/helpers/transaction.rs

```diff
@@ -643,3 +646,58 @@ pub trait LoadTransaction: SpawnBlocking {
         }
     }
 }
+
+/// Find a transaction by sender and nonce
+pub async fn get_transaction_by_sender_and_nonce<Eth>(
+    eth: &Eth,
+    sender: Address,
+    nonce: u64,
+    include_pending: bool,
+) -> Result<Option<RpcTransaction<Eth::NetworkTypes>>, Eth::Error>
+where
+    Eth: EthBlocks + EthState,
+{
+    // Check the pool first
+    if include_pending {
+        if let Some(tx) = LoadState::pool(eth).get_transaction_by_sender_and_nonce(sender, nonce) {
+            let transaction = tx.transaction.clone().into_consensus();
+            return Ok(Some(from_recovered(transaction)));
+        }
+    }
+
+    // Check if the sender is a contract
+    if EthState::get_code(eth, sender, None).await?.len() > 0 {
+        return Ok(None);
+    }
+
+    let highest = EthState::transaction_count(eth, sender, None).await?.saturating_to::<u64>();
+
+    // If the nonce is higher or equal to the highest nonce, the transaction is pending or not
+    // exists.
+    if nonce >= highest {
+        return Ok(None);
+    }
+
+    let Ok(high) = LoadBlock::provider(eth).best_block_number() else {
+        return Err(EthApiError::UnknownBlockNumber.into());
+    };
+
+    // perform a binary search over the block range to find the block in which the sender's
+    // nonce reached the requested nonce.
+    let num = binary_search::<_, _, Eth::Error>(1, high, |mid| async move {
+        let mid_nonce = EthState::transaction_count(eth, sender, Some(mid.into()))
+            .await?
+            .saturating_to::<u64>();
+
+        Ok(mid_nonce > nonce)
+    })
+    .await?;
+
+    let Some(BlockTransactions::Full(transactions)) =
+        eth.rpc_block(num.into(), true).await?.map(|block| block.transactions)
```

Comment:
no need to call `EthBlocks::rpc_block`. inspect the body of that method, you need transactions, so `LoadBlock::block_with_senders` is enough. `EthBlocks` can be replaced by `LoadBlock` as trait bound on `Eth`

Reply:
re-wrote this part here https://github.com/paradigmxyz/reth/pull/10637 since I'm using it for another pr

Reply:
thanks for the idea in #10637, get the tx in a similar way.